File: src/main/java/org/cobbzilla/util/io/CloseOnExit.java
package org.cobbzilla.util.io;

import org.slf4j.Logger;

import java.io.Closeable;
import java.util.ArrayList;
import java.util.List;

public class CloseOnExit implements Runnable {

    private static final Logger log = org.slf4j.LoggerFactory.getLogger(CloseOnExit.class);
    private static List<Closeable> closeables = new ArrayList<>();

    private CloseOnExit () {}

    static {
        Runtime.getRuntime().addShutdownHook(new Thread(new CloseOnExit()));
    }

    @Override public void run() {
        if (closeables != null) {
            for (Closeable c : closeables) {
                try {
                    c.close();
                } catch (Exception e) {
                    log.error("Error closing: " + c + ": " + e, e);
                }
            }
        }
    }

    public static void add(Closeable closeable) { closeables.add(closeable); }

}

File: src/main/java/org/cobbzilla/util/io/CloseOnExit.kt
package org.cobbzilla.util.io

import org.slf4j.Logger

import java.io.Closeable
import java.util.ArrayList

class CloseOnExit private constructor() : Runnable {

    override fun run() {
        if (closeables != null) {
            for (c in closeables) {
                try {
                    c.close()
                } catch (e: Exception) {
                    log.error("Error closing: $c: $e", e)
                }

            }
        }
    }

    companion object {

        private val log = org.slf4j.LoggerFactory.getLogger(CloseOnExit::class.java)
        private val closeables = ArrayList<Closeable>()

        init {
            Runtime.getRuntime().addShutdownHook(Thread(CloseOnExit()))
        }

        fun add(closeable: Closeable) {
            closeables.add(closeable)
        }
    }

}
